src/commands/mod.rs

```diff
@@ -2591,6 +2591,13 @@ don't make any changes, then the operation will be aborted.
         .base_repo()
         .store()
         .get_tree(&RepoPath::root(), &tree_id)?;
+    if middle_tree.id() == base_tree.id() {
+        writeln!(
+            ui.warning(),
+            "The given paths do not match any file: {}",
```

Comment:
I wish we could also give a warning if one of the paths doesn't match, but that's clearly a lot more work.

Reply:
Agreed. The `Matcher` could maybe contain a list of patterns that didn't match anything.

Reply:
Right. I was previously stuck on the fact that it'd have to be mutable for that to work. I've learned about the `Arc<Mutex>` type since. I'm not an expert, but it doesn't seem it would cause trouble here.

Reply:
I think it's better to do that by passing an additional argument parallel to the `Matcher` argument. I'd prefer to let the Matcher remain stateless [^1]. It wouldn't really know if a call to `matcher.matches("foo")` means that "foo" was actually found in the tree that was walked, although that's usually true.

[^1]: However, I have considered letting `matcher.visit("dir")` return a "subdirectory matcher" specific to visiting "dir". That might let us make it a bit more efficient by having it descendent into some internal trie, instead of having to do that for every file in "dir".